File: src/lib/quant/dataQuality.ts
import type { Bar, DataQuality } from "./types";

const TIMEFRAME_SECONDS: Record<string, number> = {
  "1Min": 60,
  "5Min": 5 * 60,
  "15Min": 15 * 60,
  "30Min": 30 * 60,
  "1Hour": 60 * 60,
  "1Day": 24 * 60 * 60,
  "1Week": 7 * 24 * 60 * 60,
};

export interface DataQualityOptions {
  asOf?: Date | string;
  minHistoryBars?: number;
  maxGapBars?: number;
  staleAfterIntervals?: number;
}

function asDate(value: Date | string): Date {
  return value instanceof Date ? value : new Date(value);
}

export function assessDataQuality(
  symbol: string,
  timeframe: string,
  bars: Bar[],
  options: DataQualityOptions = {},
): DataQuality {
  const interval = TIMEFRAME_SECONDS[timeframe] ?? null;
  const ordered = [...bars].sort((a, b) => a.t.localeCompare(b.t));
  const minHistoryBars = options.minHistoryBars ?? 60;
  const maxGapBars = options.maxGapBars ?? 3;
  const staleAfterIntervals = options.staleAfterIntervals ?? 3;

  const countMap = new Map<string, number>();
  for (const b of ordered) {
    countMap.set(b.t, (countMap.get(b.t) ?? 0) + 1);
  }
  let duplicateCount = 0;
  for (const count of countMap.values()) {
    if (count > 1) duplicateCount += count - 1;
  }
  const uniqueTimes = [...countMap.keys()].sort();

  let missingCount = 0;
  let largestGap = 0;
  if (interval && uniqueTimes.length > 1) {
    for (let i = 1; i < uniqueTimes.length; i++) {
      const prev = asDate(uniqueTimes[i - 1]);
      const current = asDate(uniqueTimes[i]);
      let gapBars = 0;
      if (timeframe === "1Day") {
        const cursor = new Date(
          Date.UTC(prev.getUTCFullYear(), prev.getUTCMonth(), prev.getUTCDate()),
        );
        cursor.setUTCDate(cursor.getUTCDate() + 1);
        while (cursor.getTime() < current.getTime()) {
          const dow = cursor.getUTCDay();
          if (dow !== 0 && dow !== 6) gapBars += 1;
          cursor.setUTCDate(cursor.getUTCDate() + 1);
        }
      } else {
        const elapsed = (current.getTime() - prev.getTime()) / 1000;
        gapBars = Math.max(0, Math.round(elapsed / interval) - 1);
      }
      missingCount += gapBars;
      largestGap = Math.max(largestGap, gapBars);
    }
  }

  const warnings: string[] = [];
  if (uniqueTimes.length < minHistoryBars) {
    warnings.push(`history_short:${uniqueTimes.length}<${minHistoryBars}`);
  }
  if (duplicateCount) {
    warnings.push(`duplicate_bars:${duplicateCount}`);
  }
  if (largestGap > maxGapBars) {
    warnings.push(`gap_too_large:${largestGap}>${maxGapBars}`);
  }

  const firstBarAt = uniqueTimes.length ? uniqueTimes[0] : null;
  const lastBarAt = uniqueTimes.length ? uniqueTimes[uniqueTimes.length - 1] : null;

  let stale = false;
  if (interval && lastBarAt) {
    const reference = options.asOf ? asDate(options.asOf) : new Date();
    const last = asDate(lastBarAt);
    const elapsedSeconds = (reference.getTime() - last.getTime()) / 1000;
    if (elapsedSeconds > interval * staleAfterIntervals) {
      stale = true;
      warnings.push("stale_last_bar");
    }
  }

  return {
    symbol: symbol.toUpperCase(),
    timeframe,
    barCount: uniqueTimes.length,
    firstBarAt,
    lastBarAt,
    expectedIntervalSeconds: interval,
    duplicateBarCount: duplicateCount,
    missingBarCount: missingCount,
    maxGapBars: largestGap,
    stale,
    isActionable: warnings.length === 0,
    warnings,
  };
}

File: backend/app/quant/data_quality.py
"""Data-quality checks for quant bars before a signal is considered actionable."""
from collections import Counter
from datetime import datetime, timedelta, timezone

from app.schemas.quant import DataQualityMetadata, OHLCVBar

TIMEFRAME_SECONDS: dict[str, int] = {
    "1Min": 60,
    "5Min": 5 * 60,
    "15Min": 15 * 60,
    "30Min": 30 * 60,
    "1Hour": 60 * 60,
    "1Day": 24 * 60 * 60,
    "1Week": 7 * 24 * 60 * 60,
}


def assess_data_quality(
    *,
    symbol: str,
    timeframe: str,
    bars: list[OHLCVBar],
    as_of: datetime | None = None,
    min_history_bars: int = 60,
    max_gap_bars: int = 3,
    stale_after_intervals: int = 3,
) -> DataQualityMetadata:
    """Return quality metadata without mutating or rejecting the supplied bars."""
    interval = TIMEFRAME_SECONDS.get(timeframe)
    ordered = sorted(bars, key=lambda bar: bar.t)
    counts = Counter(bar.t for bar in ordered)
    duplicate_count = sum(count - 1 for count in counts.values() if count > 1)
    unique = sorted(counts)

    missing_count = 0
    largest_gap = 0
    if interval and len(unique) > 1:
        for previous, current in zip(unique, unique[1:]):
            if timeframe == "1Day":
                cursor = previous
                gap_bars = 0
                while cursor.date() < current.date():
                    cursor = cursor.replace(hour=0, minute=0, second=0, microsecond=0) + timedelta(days=1)
                    if cursor.weekday() < 5 and cursor.date() < current.date():
                        gap_bars += 1
            else:
                elapsed = (current - previous).total_seconds()
                gap_bars = max(0, round(elapsed / interval) - 1)
            missing_count += gap_bars
            largest_gap = max(largest_gap, gap_bars)

    warnings: list[str] = []
    if len(unique) < min_history_bars:
        warnings.append(f"history_short:{len(unique)}<{min_history_bars}")
    if duplicate_count:
        warnings.append(f"duplicate_bars:{duplicate_count}")
    if largest_gap > max_gap_bars:
        warnings.append(f"gap_too_large:{largest_gap}>{max_gap_bars}")

    first_bar_at = ordered[0].t if ordered else None
    last_bar_at = ordered[-1].t if ordered else None
    reference_time = as_of or datetime.now(timezone.utc)
    if reference_time.tzinfo is None:
        reference_time = reference_time.replace(tzinfo=timezone.utc)
    stale = False
    if interval and last_bar_at is not None:
        last_time = last_bar_at
        if last_time.tzinfo is None:
            last_time = last_time.replace(tzinfo=timezone.utc)
        stale = (reference_time - last_time).total_seconds() > interval * stale_after_intervals
        if stale:
            warnings.append("stale_last_bar")

    return DataQualityMetadata(
        symbol=symbol.upper(),
        timeframe=timeframe,
        bar_count=len(unique),
        first_bar_at=first_bar_at,
        last_bar_at=last_bar_at,
        expected_interval_seconds=interval,
        duplicate_bar_count=duplicate_count,
        missing_bar_count=missing_count,
        max_gap_bars=largest_gap,
        stale=stale,
        is_actionable=not warnings,
        warnings=warnings,
    )
